### PR TITLE
Add support for DATA in CMake.

### PR DESCRIPTION
**Pull request by ScottTodd**
Roughly trying to mimic Bazel's behavior: https://docs.bazel.build/versions/master/build-ref.html#data. There are some differences that may require some iteration to get right.

I'm planning on using this soon for a Vulkan layer (a .so and .json file will be built/generated, and I want them accessible to `iree-run-mlir` and other binaries).

### CODE REVIEW

File: build_tools/cmake/iree_cc_binary.cmake [hunk 1]
```diff
@@ -25,6 +25,7 @@ endif()
 # Parameters:
 # NAME: name of target (see Usage below)
 # SRCS: List of source files for the binary
+# DATA: List of other targets to be built with this binary
```
Comment (GMNGeoffrey) [inline build_tools/cmake/iree_cc_binary.cmake]:
```suggestion
# DATA: List of other targets and files required for this binary
```

Same with other arg comments

File: build_tools/cmake/iree_macros.cmake [hunk 2]
```diff
@@ -123,3 +123,51 @@ function(iree_select_compiler_opts OPTS)
...
+# Data dependencies
+#-------------------------------------------------------------------------------
+
+# Adds 'data' dependencies to a target.
+#
+# Parameters:
+# NAME: name of the target to add data dependencies to
+# DATA: List of targets and/or files in the source tree. Files should use the
+#       same format as targets (i.e. iree::package::subpackage::file.txt)
+function(iree_data)
```
Comment (GMNGeoffrey) [inline build_tools/cmake/iree_macros.cmake]: Can we call this something that makes its role a little clearer? add_data_dependency?

File: build_tools/cmake/iree_macros.cmake [hunk 3]
```diff
@@ -123,3 +123,51 @@ function(iree_select_compiler_opts OPTS)
...
+      # If this file is included in multiple rules, only create the target once.
+      string(REPLACE "::" "_" _DATA_TARGET ${_DATA_LABEL})
+      if(NOT TARGET ${_DATA_TARGET})
+        set(_INPUT_PATH "${CMAKE_SOURCE_DIR}/${_FILE_PATH}")
+        set(_OUTPUT_PATH "${CMAKE_BINARY_DIR}/${_FILE_PATH}")
+        add_custom_command(
+          OUTPUT ${_OUTPUT_PATH}
+          COMMAND ${CMAKE_COMMAND} -E copy ${_INPUT_PATH} ${_OUTPUT_PATH}
+        )
+        add_custom_target(${_DATA_TARGET} DEPENDS ${_OUTPUT_PATH})
```
Comment (GMNGeoffrey) [inline build_tools/cmake/iree_macros.cmake]: Doesn't it depend on its input?

**Comment by ScottTodd** [comment]
> This would be nice to have in iree_lit_test. Have you figured out a way to test this to see if it works? CMake mostly seems to magically have whatever files except when it doesn't.

I've been iterating on this change with

```
$ rm -rf build/
$ cmake -B build/ . && cmake --build build/ --target iree_hal_vulkan_dynamic_symbols_test
$ ls build/iree/tools
(look for sanitizer_suppressions.txt)
```

Looks like we could migrate `iree_lit_test` to use this instead of `CMAKE_SOURCE_DIR`.

I also had a version of this change that put the files under more of a runfiles structure, e.g.
`build/iree/samples/simple_embedding/iree/tools/sanitizer_suppressions.txt`. The main difference there is using `CMAKE_CURRENT_BINARY_DIR` instead of `CMAKE_BINARY_DIR`. I'm less concerned about paths on host/development machines where we can walk the directory tree all we like, and more concerned with distributable builds (Android APKs, Windows exe/dll/embed within other projects, etc.). I did some reading on CMake's `install` command, but I think we can revisit that later (or it could be a good project for others to contribute to).

File: build_tools/cmake/iree_macros.cmake (hunk 3 above)
Comment (ScottTodd) [reply build_tools/cmake/iree_macros.cmake]: (Not sure I fully understand CMake's behavior here)
Indirectly, yes. I have it depending on the output path here, which is generated by the custom command using the input path. I referenced https://gitlab.kitware.com/cmake/cmake/issues/17366 when writing this.

File: build_tools/cmake/iree_macros.cmake (hunk 2 above)
Comment (ScottTodd) [reply build_tools/cmake/iree_macros.cmake]: Renamed to `iree_add_data_dependencies()`

File: build_tools/cmake/iree_cc_binary.cmake (hunk 1 above)
Comment (ScottTodd) [reply build_tools/cmake/iree_cc_binary.cmake]: 👍

**Comment by GMNGeoffrey** [comment]
> I've been iterating on this change with
> 
> ```
> $ rm -rf build/
> $ cmake -B build/ . && cmake --build build/ --target iree_hal_vulkan_dynamic_symbols_test
> $ ls build/iree/tools
> (look for sanitizer_suppressions.txt)
> ```

Cool! Yeah I've never gotten "test a target but first build all its dependencies" working right with CMake. Maybe I should try harder...

> Looks like we could migrate `iree_lit_test` to use this instead of `CMAKE_SOURCE_DIR`.
:+1: 

> I also had a version of this change that put the files under more of a runfiles structure, e.g.
> `build/iree/samples/simple_embedding/iree/tools/sanitizer_suppressions.txt`. The main difference there is using `CMAKE_CURRENT_BINARY_DIR` instead of `CMAKE_BINARY_DIR`. I'm less concerned about paths on host/development machines where we can walk the directory tree all we like, and more concerned with distributable builds (Android APKs, Windows exe/dll/embed within other projects, etc.). I did some reading on CMake's `install` command, but I think we can revisit that later (or it could be a good project for others to contribute to).

Something like a runfiles would help with https://github.com/google/iree/issues/818. I also like it from a philosophical perspective, but of course I'm steeped in Google ideas around this sort of thing :-P

File: build_tools/cmake/iree_macros.cmake (hunk 3 above)
Comment (marbre) [reply build_tools/cmake/iree_macros.cmake]: Hm, I am not sure if it is really required to call `add_custom_command()` **and** `add_custom_target`. Within `iree_py_library`, which actually tries to achieve the same (copying files to the build directory), I only call `add_custom_target`:

https://github.com/google/iree/blob/9e7cf15e7d1ce7acb2089c567346fdd7a033400c/bindings/python/build_tools/cmake/iree_py_library.cmake#L46-L49

Comment (ScottTodd) [reply build_tools/cmake/iree_macros.cmake]: 👍 that works and is simpler, thanks!

**Comment by ScottTodd** [comment]
Note: I edited the iteration steps in my comments above from `iree_hal_vulkan_dynamic_symbols_test` to `iree_hal_cts_allocator_test` (I want this for the Vulkan targets, but only affects the targets that use sanitizer_suppressions.txt right now).

**Pull request closed without merging**